README.md
# co2monitor

Automatic data logging for AirCO2ntrol mini USB co2 measurement devices.

**Hint**: This application is compatible with Raspbian.

**Note**: This application is in development state.

![co2monitor svg](https://cloud.githubusercontent.com/assets/19148271/20570532/44d42ab0-b1a4-11e6-97aa-cef5713e21cc.png)

## Background

Hendryk Plötz reverse-engineered the usb protocol on [hackaday.io](https://hackaday.io/project/5301-reverse-engineering-a-low-cost-usb-co-monitor).
I wanted to have an easy-to-use plug'n'play data logging solution for Debian/Ubuntu systems.
That's what **co2monitor** is.

## Debian package

There are ready-to-use debian packages on the [releases page](https://github.com/nobodyinperson/co2monitor/releases), you may download the [latest release](https://github.com/nobodyinperson/co2monitor/releases/latest) there.

For automatic updates, you may use my [apt repository](http://apt.nobodyinperson.de).

To build a debian package from the repository, run ```dpkg-buildpackage -us -uc``` (options mean without signing) from the repository root.
There will be a ```co2monitor_*.deb``` one folder layer above.

## Installation

Install the debian package via ```sudo dpkg -i co2monitor_*.deb```.
Older versions of **co2monitor** will automatically be removed.

If you use my [apt repository](http://apt.nobodyinperson.de), install **co2monitor** like any other package via ```sudo apt-get update && sudo apt-get install co2monitor```

Remove **co2monitor** from your system via ```sudo apt-get remove co2monitor```.

## Folder structure

The folder structure (```etc```,```lib```,```usr```,```var```) is based on the [FHS](https://en.wikipedia.org/wiki/Filesystem_Hierarchy_Standard) and will be installed as-is to the target system.

### important files
|           file                 |                  purpose              |
|--------------------------------|---------------------------------------|
| ```etc/co2monitor/service.conf``` | co2monitor service configuration file |
| ```usr/bin/co2monitor-service``` |  the co2monitor service executable |
| ```usr/bin/co2monitor``` | the co2monitor application executable |
| ```var/lib/co2monitor/data/*``` | standard folder for logged co2 data |

## Special thanks

- Hendryk Plötz on [hackaday.io](https://hackaday.io/project/5301-reverse-engineering-a-low-cost-usb-co-monitor) for the device interaction
- Mike Kazantsev on his blog on [fraggod.net](http://blog.fraggod.net/2012/06/16/proper-ish-way-to-start-long-running-systemd-service-on-udev-event-device-hotplug.html) for the systemd integration
- Ascot on [StackOverflow.com](http://stackoverflow.com/a/26457317/5433146) for a workaround on ```signal.signal(signal, handler)``` when using a ```GLib.MainLoop```
- don_crissti on [StackOveflow.com](http://unix.stackexchange.com/a/203678) for
  getting a list of dbus objects


debian/changelog
co2monitor (0.0.13) unstable; urgency=medium

  * add configurable sleep interval between readings

 -- Yann Büchau <yann.buechau@web.de>  Thu, 09 Mar 2017 09:46:59 +0100

co2monitor (0.0.12) unstable; urgency=medium

  * .desktop-file starts plot.R, not the unfinished gui

 -- Yann Büchau <yann.buechau@web.de>  Wed, 08 Mar 2017 20:03:18 +0100

co2monitor (0.0.11) unstable; urgency=medium

  * fix another 'import sys' bug... *sigh*... 
  * correct file sorting by creation time in plot.R

 -- Yann Büchau <yann.buechau@web.de>  Mon, 28 Nov 2016 14:13:21 +0100

co2monitor (0.0.10) unstable; urgency=medium

  * better hotplug handling

 -- Yann Büchau <yann.buechau@web.de>  Mon, 28 Nov 2016 11:24:14 +0100

co2monitor (0.0.9) unstable; urgency=medium

  * fix multiple co2monitor-service invocation bug

 -- Yann Büchau <yann.buechau@web.de>  Wed, 23 Nov 2016 16:09:33 +0100

co2monitor (0.0.8) unstable; urgency=medium

  * new structure: udev -> invoker -> service
  * dbus integration
  * handling of multiple devices in parallel
  * don't install some unnecessary files 
  * add French translation

 -- Yann Büchau <yann.buechau@web.de>  Wed, 16 Nov 2016 10:39:43 +0100

co2monitor (0.0.7) unstable; urgency=medium

  * better internal structure
  * use gettext for i18n and l18n
  * add German translation
  * use Glade and Gtk.Builder for gui creation
  * plug'n'play logging does now also work on non-systemd systems

 -- Yann Büchau <yann.buechau@web.de>  Fri, 04 Nov 2016 14:43:16 +0100

co2monitor (0.0.6) unstable; urgency=medium

  * add own application icon

 -- Yann Büchau <yann.buechau@web.de>  Thu, 03 Nov 2016 21:54:10 +0100

co2monitor (0.0.5) unstable; urgency=medium

  * plotting script is now auto-refreshing
  * added plotting script to application menu

 -- Yann Büchau <yann.buechau@web.de>  Wed, 02 Nov 2016 22:59:15 +0100

co2monitor (0.0.4) unstable; urgency=medium

  * better kill signal management
  * suggest r-base for plotting at installation
  * added example plotting R script to /usr/share/doc/co2monitor/examples

 -- Yann Büchau <yann.buechau@web.de>  Thu, 04 Aug 2016 10:20:36 +0200

co2monitor (0.0.3) unstable; urgency=medium

  * transition to systemd integration
  * co2monitor-service no longer killed by udevd after a while

 -- Yann Büchau <yann.buechau@web.de>  Mon, 01 Aug 2016 20:05:40 +0200

co2monitor (0.0.2) unstable; urgency=medium

  * fixed crash on boot problem
  * implemented warmup time + setting
  * included date in logs

 -- Yann Büchau <yann.buechau@web.de>  Sun, 31 Jul 2016 13:39:53 +0200

co2monitor (0.0.1) unstable; urgency=low

  * Initial release.

 -- Yann Büchau <yann.buechau@web.de>  Tue, 19 Jul 2016 11:13:58 +0200


usr/lib/co2monitor/python/co2monitor/service.py
#!/usr/bin/env python
import dbus
import dbus.service
import dbus.mainloop.glib
from xml.etree import ElementTree

import logging
import configparser 
import signal
import threading
import os, sys
import csv
import time

from gi.repository import GLib

from . import device
from . import utils

# define names
CO2MONITOR_BUSNAME    = "de.nobodyinperson.co2monitor"
CO2MONITOR_INTERFACE  = "de.nobodyinperson.co2monitor"
CO2MONITOR_OBJECTPATH = "/de/nobodyinperson/co2monitor"

###############
### Service ###
###############
class Co2MonitorService(dbus.service.Object):
    def __init__(self):
        # initially set the standard logger
        self.set_logger(logging.getLogger(__name__))
        # initially set an empty configuration
        self.set_config(configparser.ConfigParser())
        # set up the quit signals
        self.setup_signals(
            signals = [signal.SIGINT, signal.SIGTERM, signal.SIGHUP],
            handler = self.please_stop_now
        )

        # use glib as default mailoop for dbus
        dbus.mainloop.glib.DBusGMainLoop(set_as_default=True)
        dbus.mainloop.glib.threads_init()
        GLib.threads_init()

        self.loop = GLib.MainLoop() # create mainloop

        self.systembus = dbus.SystemBus() # the system bus
        self.systembus.request_name(CO2MONITOR_BUSNAME) # request the bus name
        bus_name = dbus.service.BusName(CO2MONITOR_BUSNAME, self.systembus) # create bus name
        # register the object on the bus name
        dbus.service.Object.__init__(self, bus_name, CO2MONITOR_OBJECTPATH)

    def setup_signals(self, signals, handler):
        """
        This is a workaround to signal.signal(signal, handler)
        which does not work with a GLib.MainLoop() for some reason.
        Thanks to: http://stackoverflow.com/a/26457317/5433146
        args:
            signals (list of signal.SIG... signals): the signals to connect to
            handler (function): function to be executed on these signals
        """
        def install_glib_handler(sig): # add a unix signal handler
            GLib.unix_signal_add( GLib.PRIORITY_HIGH, 
                sig, # for the given signal
                handler, # on this signal, run this function
                sig # with this argument
                )

        for sig in signals: # loop over all signals
            GLib.idle_add( # 'execute'
                install_glib_handler, sig, # add a handler for this signal
                priority = GLib.PRIORITY_HIGH  )

    # set the config
    def set_config(self, config):
        self.config = config

    # set the logger
    def set_logger(self, logger):
        self.logger = logger

    # kindly stop logging
    def please_stop_now(self, signum=None, frame=None):
        self.logger.info(" ".join([_("received stop signal {},"),
            _("will terminate soon...")]).format(signum))
        self.quit()

    def run(self):
        self.logger.info(_("Service running..."))
        self.loop.run()
        self.logger.info(_("Service stopped"))

    # get a list of exported objects under a path (only one layer deep)
    # thanks to don_crissti on http://unix.stackexchange.com/a/203678
    def get_monitored_devices_objects(self):
        res = []
        # iface = dbus.Interface(self, 'org.freedesktop.DBus.Introspectable')
        xml_string = self.Introspect(CO2MONITOR_OBJECTPATH,self.connection)
        for child in ElementTree.fromstring(xml_string):
            if child.tag == 'node':
                new_path = '/'.join((CO2MONITOR_OBJECTPATH, child.attrib['name']))
                res.append(new_path)
        return res

    @dbus.service.method(CO2MONITOR_INTERFACE, 
        in_signature='s', out_signature='b')
    def start_device_logging(self, devicefile):
        self.logger.info(_("received request to start logging on device {}").format(
            devicefile))
        # create the object path name on the bus name
        objectpath = "/".join([CO2MONITOR_OBJECTPATH,
            utils.devicefile2objectname(devicefile)])
        # check if device is already monitored
        monitored_devices = self.get_monitored_devices_objects()
        if objectpath in monitored_devices:
            self.logger.warning(" ".join([
            _("There is already a logging thread for device {}."),
            _("This is odd... I better do nothing.")
            ]).format(devicefile))
            return False
        else:
            try:
                thread = LogThread(devicefile = devicefile) # logger object
                self.logger.info(_("starting logging thread for device {}").format(
                    devicefile))
                # same logger as service
                thread.set_logger(self.logger)
                # same config as service
                thread.set_config(self.config)
                thread.daemon = True # let this thread be a daemon thread
                thread.start()
            except OSError:
                self.logger.critical(_("Could not access the device file '{}'."
                    ).format(devicefile))
                return False
            except:
                self.logger.critical(_(
                    "Something went wrong with device file '{}'."
                    ).format(devicefile))
                return False
            return True


    @dbus.service.method(CO2MONITOR_INTERFACE, in_signature='', out_signature='s')
    def status(self):
        if self.loop.is_running():
            return(_("running"))
        else:
            return(_("stopped"))
            

    @dbus.service.method(CO2MONITOR_INTERFACE, in_signature='', out_signature='')
    def quit(self):
        self.logger.info(_("stopping co2monitor..."))
        self.loop.quit()
        self.logger.info(_("stopped co2monitor"))


#################
### LogThread ###
#################
class LogThread(threading.Thread, dbus.service.Object):
    def __init__(self, devicefile):
        # by default, log!
        self.do_data_logging = True
        # initially set the standard logger
        self.set_logger(logging.getLogger(__name__))
        # initially set an empty configuration
        self.set_config(configparser.ConfigParser())

        # use glib as default mailoop for dbus
        dbus.mainloop.glib.DBusGMainLoop(set_as_default=True)
        dbus.mainloop.glib.threads_init()
        GLib.threads_init()

        self.systembus = dbus.SystemBus() # the system bus
        self.systembus.request_name(CO2MONITOR_BUSNAME) # request the bus name
        bus_name = dbus.service.BusName(CO2MONITOR_BUSNAME, self.systembus) # create bus name

        self.devicefile = devicefile

        # set up the device
        self.device = device.co2device(self.devicefile)

        # register the object on the bus name
        objectpath = "/".join([CO2MONITOR_OBJECTPATH,
            utils.devicefile2objectname(self.devicefile)])
        dbus.service.Object.__init__(self, bus_name, objectpath)
        self.update_status(_("idle"))
        threading.Thread.__init__(self)


    # set the config
    def set_config(self, config):
        self.config = config

    # set the logger
    def set_logger(self, logger):
        self.logger = logger

    @dbus.service.method(CO2MONITOR_INTERFACE, in_signature='', out_signature='s')
    def status(self):
        return self.current_status

    def update_temperature(self, temperature):
        self.current_temperature = temperature

    @dbus.service.method(CO2MONITOR_INTERFACE, in_signature='', out_signature='d')
    def temperature(self):
        try:    temperature = float(self.current_temperature) # try to convert to float
        except: temperature =           -99999                # fake measurement
        return temperature

    def update_co2(self, co2):
        self.current_co2 = co2

    @dbus.service.method(CO2MONITOR_INTERFACE, in_signature='', out_signature='i')
    def co2(self):
        try:    co2 = int(self.current_co2) # try to convert to int
        except: co2 = -99999                # fake measurement
        return co2

    # kindly stop logging
    @dbus.service.method(CO2MONITOR_INTERFACE, in_signature='', out_signature='')
    def please_stop_now(self):
        self.logger.debug(_("Request to stop logging on device {} - will terminate soon...").format(self.devicefile))
        self.do_data_logging = False

    # update the current status
    def update_status(self, status):
        self.current_status = status

    # shut this thread down
    def quit(self):
        self.logger.debug(_("Shutting down logging thread of device {}").format(
            self.devicefile))
        self.remove_from_connection()
        self.logger.debug(_("logging thread for device {} was shut down properly.").format(
            self.devicefile))
        sys.exit()

    # logloop
    def run(self):
        self.logger.info(_("Starting data logging on device {}").format(
            self.devicefile))
        datafolder = self.config.get('data-logging','datafolder')

        self.update_status(_("creating data directory"))
        if not os.path.isdir(datafolder): # check if data folder does not exist
            os.makedirs(datafolder,0o755) # create data folder
        self.update_status(_("created data directory"))
        
        fieldnames = ['time','temperature','co2']

        datafile   = "{folder}/co2monitor-{device}-{time}.csv".format(
            time = time.strftime("%Y%m%d%H%M%S",time.localtime()),
            device = os.path.basename(self.devicefile),
            folder = datafolder
            )

        # warmup
        warmuptime = self.config.getint('data-logging','warmuptime')
        waittime = round(max(0,warmuptime - self.device.uptime()))
        if waittime > 0:
            self.update_status(_("device warmup"))
            self.logger.info(
                _("giving the device another {} seconds of warmup time...").format(
                waittime))
            for i in range(waittime): 
                if not self.do_data_logging: break # stop if asked to stop
                time.sleep(1) # sleep one second

        self.update_status(_("opening data file"))
        with open(datafile, 'w') as csvfile: # open file
            self.update_status(_("opened data file"))
            self.logger.info(_("opened {} for data logging.").format(datafile))

            writer = csv.DictWriter(csvfile, fieldnames = fieldnames)

            self.update_status(_("writing header to file"))
            writer.writeheader()
            measold = {}
            while self.do_data_logging: # read as long as you are allowed
                # self.logger.debug(_("reading from device {}...").format(self.devicefile))
                self.update_status(_("reading from device"))
                meas = self.device.read() # read from device
                # self.logger.debug(_("read {} from device {}").format(meas,self.devicefile))

                self.update_status(_("postprocessing raw data"))
                if isinstance(meas, dict): # measurement worked
                    # set time
                    meas['time'] = time.strftime("%Y-%m-%d %H:%M:%S",
                                                 time.localtime())
    
                    anything_new = False

                    temperature     =    meas.get('temperature',None)
                    temperature_old = measold.get('temperature',None)
                    co2     =    meas.get('co2',None)
                    co2_old = measold.get('co2',None)

                    if temperature != temperature_old:
                        self.update_temperature(temperature)
                        anything_new = True
                    if co2 != co2_old:
                        self.update_co2(co2)
                        anything_new = True
                    # check if this measurement introduces something new
                    if anything_new:
                        # measurement is not only None
                        if not temperature is None or not co2 is None:
                            self.update_status(_("writing data to file"))
                            writer.writerow(meas) # write
                            csvfile.flush() # writeout
    
                    
                    # update old measurements
                    measold.update(meas)
                else: # measurement didn't work
                    break # exit logloop
                
                # wait predefined number of seconds
                try: 
                    waittime = int(
                        self.config["data-logging"].getint("interval"))
                    assert waittime > 0
                except: 
                    waittime = 0
                time.sleep(waittime)

            # logging ended
            self.logger.info(_("stopped logging."))
        self.logger.debug(_("closed data logging file '{}'").format(datafile))

        self.logger.info(_("Stopped data logging on device {}").format(
            self.devicefile))
        # quit
        self.quit() 
            

